### PR TITLE
Changes to container runlabel for toolbox project

### PR DESCRIPTION
**Pull request by baude**
The toolbox project would benefit from a few changes to more closely
resembe the original atomic cli project.  Changes made are:

* only pull image for container runlabel if the label exists in the image
* if a container image does not have the desired label, exit with non-zero

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM

File: libpod/adapter/runtime_remote.go [hunk 1]
```diff
@@ -156,7 +156,7 @@ func (r *LocalRuntime) LoadFromArchiveReference(ctx context.Context, srcRef type
 }
 
 // New calls into local storage to look for an image in local storage or to pull it
-func (r *LocalRuntime) New(ctx context.Context, name, signaturePolicyPath, authfile string, writer io.Writer, dockeroptions *image.DockerRegistryOptions, signingoptions image.SigningOptions, forcePull bool) (*ContainerImage, error) {
+func (r *LocalRuntime) New(ctx context.Context, name, signaturePolicyPath, authfile string, writer io.Writer, dockeroptions *image.DockerRegistryOptions, signingoptions image.SigningOptions, forcePull bool, label *string) (*ContainerImage, error) {
```
Comment (mtrmac) [inline libpod/adapter/runtime_remote.go]: Why is it OK to just silently ignore the `label` parameter?

File: cmd/podman/shared/container.go [hunk 2]
```diff
@@ -609,7 +609,7 @@ func GetRunlabel(label string, runlabelImage string, ctx context.Context, runtim
 			registryCreds = creds
 		}
 		dockerRegistryOptions.DockerRegistryCreds = registryCreds
-		newImage, err = runtime.ImageRuntime().New(ctx, runlabelImage, signaturePolicyPath, authfile, output, &dockerRegistryOptions, image.SigningOptions{}, false)
+		newImage, err = runtime.ImageRuntime().New(ctx, runlabelImage, signaturePolicyPath, authfile, output, &dockerRegistryOptions, image.SigningOptions{}, false, &label)
```
Comment (mtrmac) [inline cmd/podman/shared/container.go]: Hum… so, ultimately, this code does not need the image layers pulled _at all_, and all it only needs the labels? That’s fairly wasteful (though it’s not easy to get the `pullGoal` code in any other way, right now).

Or would the image get pulled either way?

File: libpod/image/pull.go [hunk 3]
```diff
@@ -218,27 +219,54 @@ func (ir *Runtime) pullImageFromReference(ctx context.Context, srcRef types.Imag
...
+
 	for _, imageInfo := range goal.refPairs {
 		copyOptions := getCopyOptions(sc, writer, dockerOptions, nil, signingOptions, "", nil)
 		copyOptions.SourceCtx.SystemRegistriesConfPath = systemRegistriesConfPath // FIXME: Set this more globally.  Probably no reason not to have it in every types.SystemContext, and to compute the value just once in one place.
 		// Print the following statement only when pulling from a docker or atomic registry
 		if writer != nil && (imageInfo.srcRef.Transport().Name() == DockerTransport || imageInfo.srcRef.Transport().Name() == AtomicTransport) {
 			io.WriteString(writer, fmt.Sprintf("Trying to pull %s...", imageInfo.image))
 		}
+		// If the label is not nil, check if the label exists and if not, return err
+		if label != nil {
```
Comment (mtrmac) [inline libpod/image/pull.go]: (Please consider moving all of this into a separate helper function, to make the core pull loop readable without overwhelming readers with this corner case.)

Comment (baude) [reply libpod/image/pull.go]: agreed, done

**Comment by baude** [comment]
addresses @mtrmac comments and also added two tests for regression.

File: cmd/podman/shared/container.go (hunk 2 above)
Comment (mtrmac) [reply cmd/podman/shared/container.go]: > if the label doesn't exist in the remote image, we don't pull it because the image would be useless.

Sure, if the label does not exist; but if the label _does_ exist, will the image get pulled soon after, or is that possible to be a big waste?

**Comment by mtrmac** [comment]
> The code looks good to me but I don't like adding the `label` parameter. It seems like a corner case that determines the API. I would prefer to make `checkRemoteImageForLabel` public and call it before pulling the image; or just copy the code.

The tricky part about is is all the heuristics involved in converting the user input into a `types.ImageReference`; we really don’t want to invent a new different set of heuristics.

It would be possible to make the `pullGoal` determination a public function, or move the “get and check label” code into the `image` subpackage.

What is best, I think, depends on whether this is really a check within an existing pull process, or a basically standalone read/inspect that does not typically involve pulling an entire image.

> Besides the pure design arguments, I would only change libpod's API in extreme cases to allow potential users to update without having API breakage.

(Does libpod have an API stability goal? I have no idea, so far I was relying on the repo owners to handle that.)

**Comment by vrothberg** [comment]
> What is best, I think, depends on whether this is really a check within an existing pull process, or a basically standalone read/inspect that does not typically involve pulling an entire image.

In my opinion, that should be a standalone check to avoid breaking the API. If API breakage is acceptable, I strongly encourage to encapsulate the label into a more generic struct (e.g., a `type ImageProperties struct`) to avoid adding more parameters in the future.

> (Does libpod have an API stability goal? I have no idea, so far I was relying on the repo owners to handle that.)

I don't know of a public statement :) But I know of tools using libpod directly instead of shelling out to Podman (wrote one at my former employer).

**Comment by mtrmac** [comment]
> In my opinion, that should be a standalone check to avoid breaking the API. If API breakage is acceptable, I strongly encourage to encapsulate the label into a more generic struct (e.g., a `type ImageProperties struct`) to avoid adding more parameters in the future.

Buildah has `PullOptions` FWIW; if we have such a generic `New` method, being explicit about what the options are for could be nice. Or maybe we do want a very generic `NewOptions` so that the API needs to never change? I don’t know, on balance I’d rather make it difficult to stuff too many very different things in there.

BTW not breaking the API does not inherently require the check to be stand-alone (if that meant low-performing): we can introduce a `NewWithOptions`, and keep the existing `New` as a compatibility wrapper that calls `NewWithOptions(…, {})`.

**Comment by baude** [comment]
I  initially had it being separate but as @mtrmac says, the heuristics are proven to be correct right here.  I would be fine with breaking it out again so lets make a decision.  This needs to get in for the toolbox folks.

File: cmd/podman/shared/container.go (hunk 2 above)
Comment (mtrmac) [reply cmd/podman/shared/container.go]: Per IRC, the image would get pulled if the label exists.

**Comment by rhatdan** [comment]
Don't we have a mechanism to pull down the entire json file that you see with podmain inspect,  I believe is what skopeo inspect is supposed to do.  In this case podman should pull down the image.json file and look for whatever labels he cares about.

**Comment by baude** [comment]
before this bitrots,can we merge or suggest a realistic alternative?

File: libpod/adapter/runtime_remote.go (hunk 1 above)
Comment (rhatdan) [reply libpod/adapter/runtime_remote.go]: @baude Did you switch to returning an error?

Comment (baude) [reply libpod/adapter/runtime_remote.go]: yes

**Comment by mtrmac** [comment]
LGTM.

**Comment by rhatdan** [comment]
This one also seems to have hung up.
